### DISCUSSION TITLE
Don't necessarily tear down RTSP stream

### DISCUSSION THREAD

**Discussion by BruceHowells** [tsightler/ring-mqtt#355] [Feature Requests]
I was hoping to do some "offboard" analysis of events by leveraging the RTSP support.  Specifically, for a first swag, I'm trying to have Zoneminder watch the RTSP event stream.  (This works with the _live stream, but I don't want to do that for many reasons already in the documentation.)

Unfortunately, at the end of the playback of the event, the stream drops, which makes Zoneminder reconnect, which retrieves the last event until it stops, and then...

In a fantasy, I'd love for the option to, rather than drop the RTSP connection, instead substitute some kind of idle feed (a simple black frame, perhaps) and leave the stream up.  This would let consumers stay connected without reconnect loops, and the logic to detect "stream gone...  ok, stream back" would be very simple for them to deal with as well.

Any alternatives to the problem of "new connect?  here's the last event.  Ok, go away now..  oh, your're back, here's the last event." would be very welcome as well.  I was trying to brainstorm a way to RTSP (or other) stream based on new files appearing in a directory, but I've not seen anything like that around, either.

**Comment by tsightler** [comment]
I've thought in the past about the idea of having a "continuous" stream for both the live and event stream, but this is much more complex.  The current model is super simple since it's based on the built-in capability in rtsp-simple-server to run a script to start a stream only when a client is connected.  If the stream was always publishing there'd be no method to know when to trigger the actual stream to start, because the stream would already have a publisher.  It might be possible to poll the API to see if there are clients, but this is pretty ugly and it just adds a lot of complication.

Personally, my suggestion is to buy a camera that is designed for continuous streaming that will work well with Zoneminder vs trying to figure out some way to make a cloud based smart camera try do what you want because it will always be a sub-par experience, IMO.

However, if you're intent on trying to get the Ring camera to do this I'm pretty sure there's already some good info on doing basically this on the Zoneminder wiki using an mp4 file as a source.  Also, I'm pretty sure you can use the file source which I believe uses low frame rate jpeg images, so a simple script to download and playback the file as a low frame rate jpeg overwritting the same file over and over would probably do the trick, but I've admittedly not used Zoneminder in quite a while.  Regardless, I believe that is the best and most reliable path, i.e. download the file and then do something with it.

**Reply by tsightler** [reply]
Here's an older entry on Zoneminder wiki to use a video file as a source:  https://wiki.zoneminder.com/How_to_use_a_video_file_as_a_Source

**Comment by BruceHowells** [comment]
I always knew it would be sub-optimal, but I'm balancing that with the fact
that I've got a bunch of Ring devices out there already and was trying not
to have my front yard look like something out of 1984.  :)  Thanks for the
reply and thought on this, and that ZM link - that seems to solve my "I've
got a file, how do I get it into ZM" problem and will keep me busy for a
while.

**Reply by tsightler** [reply]
Sounds good, if you get it working I think it would be something that would be really interesting to share as the question does come up from time to time.

**Comment by sts098** [comment]
Wondering if there has been any progress on the ZoneMinder type continuous stream option here.  I also have a number of ring devices, and while I am open to continuous type doorbells like Amcrest, I am not looking forward to fishing all the wires as I currently run my rings on batteries.

**Reply by tsightler** [reply]
No progress from ring-mqtt perspective, it's just too complex for me right now since I have very little time and don't know enough about video streaming and muxing multiple streams together.

I do believe the problem from the OP in this thread, which wanted to do "offboard" analysis of streams, could be fairly easily solved with a little creative automation and API use, toggling the ZM monitor between "None" and "Always" capture mode via the API would seem doable.

The question to me is, how would you trigger analysis to begin with?  I mean, streaming all the time will kill the battery, so that's not a good option unless you already have wired cameras.  If you could trigger ZM to analyze the video only during motion events, since those are already recording, would that be good enough for your use case?  What does ZM add in this case that Ring does have?  Better object detection?  Just curious, I really don't know much about ZM.  Thanks for sharing?

**Reply by sts098** [reply]
Thanks for the reply.  I will try to explain what I had hoped to accomplish.  

My Configuration:
- First, I do not wish to maintain an active subscription to Ring Monitoring.
- I am running Home Assistant
- I do have a QNAP running QVR Pro (NVR) 
- various cameras are recorder
- I have installed ring-MQTT (which is great) and I can display Streams in Home Assistant
- I do not believe I can record streams to my NVR as from my understanding, as long as my NVR remains connected, Simple RTSP will keep the stream open, continuously streaming and trying to maintain connection to ring.  

What I was hoping to do is:
- monitor Ding and Possibly Motion
- Turn on Live Stream for a timed period
- Turn off Live Stream

Hoping to serve a NEW RTSP Stream from Simple RTSP Server that would carry the Ring Camera Stream during Live Stream ON and a blank screen or image when live Stream Switch is OFF.  

Would hope to connect the NVR to this NEW Stream, which could stay connected to the NVR and not cause the Ring Camera Stream to stream continuously.

**Reply by tsightler** [reply]
OK, well, that's not something I would be interested in adding to this project.  I think it could be done, but as your primary goal is to **not** use Ring services, and that is not the goal of this project, then I can safely say that's not something I'm interested in pursuing.  Thanks!

**Reply by sts098** [reply]
Thanks anyway!  Great Project!

**Comment by BruceHowells** [comment]
To hijack the thread a little…. I’m currently using Home Assistant to grab videos from Ring to local storage, but I’m struggling with how to wake up ZoneMinder or something similar to ingest the files.
Thinking about it, maybe some kind of outbound web hook would be helpful, and maybe I’ll take another swing at it soon now that I’ve survived the Home Assistant Great Bluetooth Refactoring.
My original approach was very much like what the original poster reported, but I realized that was either wrong or infeasible.  Looks like I got that right.
This is also likely well out of scope for ring-mqtt, but thought I’d waste an email in case someone is following the same path in life.

**Reply by tsightler** [reply]
Did you try using an automation to enable/disable the monitor in ZoneMinder via the API?  While I've never used ZM, the docs make this look pretty straightforward.  I feel like you could just set up a ZoneMinder monitor with RTSP pointed at the live or event stream, then use an automation to run a script with curl to toggle the particular monitor from None to Always and then back after some time.  Seems pretty simple and straightforward to me.  If I had time I'd play with it myself.

**Comment by BruceHowells** [comment]
That doesn’t directly abandon my RTSP thoughts. Hm. If I get it to work, I’ll share the experience if you wish to stuff it in a FAQ or something. On Oct 19, 2022, at 2:36 PM, tsightler ***@***.***> wrote:﻿
Did you try using an automation to enable/disable the monitor in ZoneMinder via the API?  While I've never used ZM, the docs make this look pretty straightforward.  I feel like you could just set up a ZoneMinder monitor with RTSP pointed at the live or event stream, then use an automation to run a script with curl to toggle the particular monitor from None to Always and then back after some time.  Seems pretty simple and straightforward to me.  If I had time I'd play with it myself.

**Reply by tsightler** [reply]
Correct, the general idea is that you can just use a simple "Start monitor" and "Stop monitor" script that just uses curl to POST to the API to start and stop the monitor based on a motion event.  Looking at the API something like this should be able to start the monitor analyzing and, optionally, recording (triggered on a motion event from the camera):
```
curl -XPOST http://server/zm/api/monitors/1.json -d 
"Monitor[Capturing]=Always&Monitor[Analysing]=Always&Monitor[Recording]=Always"
```
And then, after X number of seconds the monitor would be disabled with:
```
curl -XPOST http://server/zm/api/monitors/1.json -d 
"Monitor[Capturing]=None&Monitor[Analysing]=None&Monitor[Recording]=None"
```
You wouldn't need any extra RTSP streams or anything like that in ring-mqtt, just run automation on motion to start the monitor and then stop the monitor after X seconds which would automatically shut down the RTSP stream.  To me this should be completely doable, but I may be misunderstanding something about the way ZoneMinder works as I'm basing it 100% on documentation and zero hands on knowledge.
